File: robot_gazebo/src/ground_truth_publisher.cpp
/**
 * Ground Truth Publisher (C++)
 * ============================
 * Subscribes directly to Gazebo's /world/default/dynamic_pose/info transport
 * topic via gz::transport (bypasses ros_gz_bridge name-mapping bug).
 *
 * Reads gz::msgs::Pose_V, finds the entry whose .name() == robot_name param,
 * and publishes:
 *   /ground_truth/odom  (nav_msgs/msg/Odometry)
 *   /ground_truth/pose  (geometry_msgs/msg/PoseStamped)
 *
 * Why C++ instead of Python: gz-python transport bindings are not installed
 * and the ros_gz_bridge Pose_V→TFMessage conversion discards model names
 * (child_frame_id ends up empty).
 */

#include <string>
#include <memory>
#include <chrono>
#include <cmath>

#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>

#include <gz/transport/Node.hh>
#include <gz/msgs/pose_v.pb.h>

using namespace std::chrono_literals;

class GroundTruthPublisher : public rclcpp::Node
{
public:
  GroundTruthPublisher()
  : Node("ground_truth_publisher"),
    msg_count_(0)
  {
    // ── Parameters ─────────────────────────────────────────────────────────
    this->declare_parameter<std::string>("robot_name", "robot");
    this->declare_parameter<std::string>("world_name", "default");
    this->declare_parameter<std::string>("map_frame", "map");
    this->declare_parameter<std::string>("ground_truth_child_frame", "ground_truth_base_link");
    // Offset from Gazebo world frame to the ROS map frame (translation + yaw).
    // The SLAM map frame origin is wherever the robot was in the Gazebo world
    // when the mapping session started.  If that starting position was NOT at
    // Gazebo world (0, 0), the two frames differ by exactly that offset.
    // Set these to the robot's Gazebo world (x, y) position at SLAM-start
    // (equivalently: observe where /fgo/odometry settles when the robot is at
    // Gazebo world origin after NDT locks on; that reported position is the
    // offset you need here, with opposite sign).
    this->declare_parameter<double>("world_to_map_offset_x",   0.0);
    this->declare_parameter<double>("world_to_map_offset_y",   0.0);
    this->declare_parameter<double>("world_to_map_offset_yaw", 0.0);

    robot_name_   = this->get_parameter("robot_name").as_string();
    world_name_   = this->get_parameter("world_name").as_string();
    map_frame_    = this->get_parameter("map_frame").as_string();
    child_frame_  = this->get_parameter("ground_truth_child_frame").as_string();
    off_x_   = this->get_parameter("world_to_map_offset_x").as_double();
    off_y_   = this->get_parameter("world_to_map_offset_y").as_double();
    off_yaw_ = this->get_parameter("world_to_map_offset_yaw").as_double();

    // ── ROS publishers ──────────────────────────────────────────────────────
    odom_pub_ = this->create_publisher<nav_msgs::msg::Odometry>(
      "/ground_truth/odom", 10);
    pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>(
      "/ground_truth/pose", 10);

    // ── gz::transport subscriber ────────────────────────────────────────────
    // Direct connection to Gazebo physics engine – zero sensor noise.
    std::string gz_topic =
      "/world/" + world_name_ + "/dynamic_pose/info";

    bool ok = gz_node_.Subscribe(
      gz_topic,
      &GroundTruthPublisher::onGzPose,
      this);

    if (!ok) {
      RCLCPP_ERROR(
        this->get_logger(),
        "Failed to subscribe to gz topic: %s", gz_topic.c_str());
    } else {
      RCLCPP_INFO(
        this->get_logger(),
        "GroundTruthPublisher started\n"
        "  Gz topic   : %s\n"
        "  Robot name : %s\n"
        "  Publishing : /ground_truth/odom  (nav_msgs/Odometry)\n"
        "  Publishing : /ground_truth/pose  (geometry_msgs/PoseStamped)",
        gz_topic.c_str(), robot_name_.c_str());
    }

    // Watchdog: warn every 5 s if no data received
    watchdog_ = this->create_wall_timer(
      5s, [this]() {
        if (msg_count_ == 0) {
          RCLCPP_WARN(
            this->get_logger(),
            "No ground truth poses received yet. "
            "Check that Gazebo is running and model name matches '%s'.",
            robot_name_.c_str());
        } else {
          RCLCPP_INFO(
            this->get_logger(),
            "[ground_truth] Publishing OK – %lu msgs in last 5 s.", msg_count_);
          msg_count_ = 0;
        }
      });
  }

private:
  // ── gz callback (called from gz transport thread) ─────────────────────────
  void onGzPose(const gz::msgs::Pose_V & poses_msg)
  {
    for (int i = 0; i < poses_msg.pose_size(); ++i) {
      const auto & pose = poses_msg.pose(i);

      if (pose.name() != robot_name_) {
        continue;
      }

      // Timestamp from Gazebo header (simulation time)
      builtin_interfaces::msg::Time stamp;
      stamp.sec     = poses_msg.header().stamp().sec();
      stamp.nanosec = poses_msg.header().stamp().nsec();

      const auto & pos = pose.position();
      const auto & ori = pose.orientation();

      // ── nav_msgs/Odometry ─────────────────────────────────────────────────
      nav_msgs::msg::Odometry odom;
      odom.header.stamp    = stamp;
      odom.header.frame_id = map_frame_;
      odom.child_frame_id  = child_frame_;

      // Apply world→map frame offset so the published pose is in the same
      // coordinate system as /fgo/odometry (the ROS map frame).
      // T_map_robot = T_world_map^{-1} * T_world_robot
      // For a pure 2-D offset (tx, ty, yaw) the inverse is: (-tx*cos(yaw)-ty*sin(yaw),
      // tx*sin(yaw)-ty*cos(yaw), -yaw), but since Gazebo world and map frame have the
      // same Z-up orientation (yaw offset = 0 by convention for a correctly built map),
      // this simplifies to a plain XY translation:
      //   x_map = cos(off_yaw)*(x_world - off_x) + sin(off_yaw)*(y_world - off_y)
      //   y_map = -sin(off_yaw)*(x_world - off_x) + cos(off_yaw)*(y_world - off_y)
      const double cos_yaw = std::cos(off_yaw_);
      const double sin_yaw = std::sin(off_yaw_);
      const double dx = pos.x() - off_x_;
      const double dy = pos.y() - off_y_;
      const double map_x =  cos_yaw * dx + sin_yaw * dy;
      const double map_y = -sin_yaw * dx + cos_yaw * dy;

      // Yaw in world frame, corrected by map offset yaw
      // (Extract yaw from quaternion, subtract offset, rebuild quaternion)
      const double gz_qw = ori.w(), gz_qx = ori.x(), gz_qy = ori.y(), gz_qz = ori.z();
      const double world_yaw = std::atan2(
        2.0 * (gz_qw * gz_qz + gz_qx * gz_qy),
        1.0 - 2.0 * (gz_qy * gz_qy + gz_qz * gz_qz));
      const double map_yaw = world_yaw - off_yaw_;
      const double half = map_yaw * 0.5;
      const double map_qw =  std::cos(half);
      const double map_qz =  std::sin(half);

      odom.pose.pose.position.x    = map_x;
      odom.pose.pose.position.y    = map_y;
      odom.pose.pose.position.z    = pos.z();
      odom.pose.pose.orientation.x = 0.0;
      odom.pose.pose.orientation.y = 0.0;
      odom.pose.pose.orientation.z = map_qz;
      odom.pose.pose.orientation.w = map_qw;
      // All covariances are zero → perfect ground truth
      odom.pose.covariance.fill(0.0);
      odom.twist.covariance.fill(0.0);

      odom_pub_->publish(odom);

      // ── geometry_msgs/PoseStamped ─────────────────────────────────────────
      geometry_msgs::msg::PoseStamped ps;
      ps.header.stamp    = stamp;
      ps.header.frame_id = map_frame_;
      ps.pose            = odom.pose.pose;

      pose_pub_->publish(ps);

      ++msg_count_;
      return;  // found the robot, no need to iterate further
    }
  }

  // ── Members ───────────────────────────────────────────────────────────────
  std::string robot_name_;
  std::string world_name_;
  std::string map_frame_;
  std::string child_frame_;
  double off_x_{};
  double off_y_{};
  double off_yaw_{};

  gz::transport::Node gz_node_;

  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pose_pub_;

  rclcpp::TimerBase::SharedPtr watchdog_;
  std::size_t msg_count_;
};


int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<GroundTruthPublisher>());
  rclcpp::shutdown();
  return 0;
}

File: robot_gazebo/scripts/ground_truth_publisher.py
#!/usr/bin/env python3
"""
Ground Truth Publisher for Localization Benchmarking
=====================================================
Extracts the robot's true pose directly from the Gazebo physics engine
via the /world/default/dynamic_pose/info bridge (gz.msgs.Pose_V → TFMessage).

Published Topics:
  /ground_truth/odom  (nav_msgs/Odometry)      – for algorithm comparison
  /ground_truth/pose  (geometry_msgs/PoseStamped) – for RViz overlay

The "world" frame in Gazebo ≡ "map" frame in ROS 2 (fixed inertial origin).
No sensor noise, no drift – 100% accurate position from the simulator.
"""

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSDurabilityPolicy

from tf2_msgs.msg import TFMessage
from nav_msgs.msg import Odometry
from geometry_msgs.msg import (
    PoseStamped,
    TransformStamped,
    Quaternion,
    Point,
    Vector3,
)
from builtin_interfaces.msg import Time

# The Gazebo model name as declared in model.sdf: <model name='robot'>
ROBOT_MODEL_NAME = "robot"

# Gazebo "world" frame maps to ROS "map" frame (inertial, fixed origin)
MAP_FRAME = "map"
GROUND_TRUTH_FRAME = "ground_truth_base_link"


class GroundTruthPublisher(Node):
    """
    Bridges the Gazebo dynamic pose vector to clean ROS 2 ground truth topics.

    The gz→ROS bridge publishes every model's world-frame pose as a
    TFMessage on /gz/dynamic_pose.  This node filters the robot entry and
    re-publishes it in formats convenient for localization comparison
    (nav_msgs/Odometry and geometry_msgs/PoseStamped).
    """

    def __init__(self) -> None:
        super().__init__("ground_truth_publisher")

        # ── Parameters ─────────────────────────────────────────────────────
        self.declare_parameter("robot_name", ROBOT_MODEL_NAME)
        self.declare_parameter("map_frame", MAP_FRAME)
        self.declare_parameter("ground_truth_child_frame", GROUND_TRUTH_FRAME)
        self.declare_parameter("publish_rate_warn_hz", 0.2)  # log warn if silent > 5 s

        self._robot_name: str = (
            self.get_parameter("robot_name").get_parameter_value().string_value
        )
        self._map_frame: str = (
            self.get_parameter("map_frame").get_parameter_value().string_value
        )
        self._child_frame: str = (
            self.get_parameter("ground_truth_child_frame")
            .get_parameter_value()
            .string_value
        )

        # ── Subscribers ────────────────────────────────────────────────────
        # /gz/dynamic_pose is the remapped bridge topic for
        # /world/default/dynamic_pose/info (gz.msgs.Pose_V → TFMessage)
        self._tf_sub = self.create_subscription(
            TFMessage,
            "/gz/dynamic_pose",
            self._on_gz_tf,
            qos_profile=10,
        )

        # ── Publishers ─────────────────────────────────────────────────────
        # nav_msgs/Odometry – standard format for localization comparison
        self._odom_pub = self.create_publisher(
            Odometry,
            "/ground_truth/odom",
            qos_profile=10,
        )

        # geometry_msgs/PoseStamped – easy to plot in RViz as an arrow/path
        self._pose_pub = self.create_publisher(
            PoseStamped,
            "/ground_truth/pose",
            qos_profile=10,
        )

        # ── State ──────────────────────────────────────────────────────────
        self._msg_count: int = 0
        self._watchdog = self.create_timer(5.0, self._watchdog_cb)
        self._last_stamp: Time | None = None

        self.get_logger().info(
            f"GroundTruthPublisher started — tracking model '{self._robot_name}'\n"
            f"  Subscribing : /gz/dynamic_pose  (TFMessage)\n"
            f"  Publishing  : /ground_truth/odom  (nav_msgs/Odometry)\n"
            f"  Publishing  : /ground_truth/pose  (geometry_msgs/PoseStamped)\n"
            f"  Map frame   : {self._map_frame}"
        )

    # ── Callbacks ───────────────────────────────────────────────────────────

    def _on_gz_tf(self, msg: TFMessage) -> None:
        """
        Called for every TFMessage from the Gazebo dynamic pose bridge.
        The message contains one TransformStamped per dynamic model/link.
        We search for the entry whose child_frame_id matches our robot model.
        """
        robot_tf: TransformStamped | None = None

        for transform in msg.transforms:
            # Gazebo Harmonic names the Pose_V entries after the model name
            # (not individual links).  The child_frame_id will be "robot".
            if transform.child_frame_id == self._robot_name:
                robot_tf = transform
                break

        if robot_tf is None:
            return  # robot not yet spawned or name mismatch

        self._msg_count += 1
        stamp = robot_tf.header.stamp

        # ── Publish Odometry ────────────────────────────────────────────────
        odom = Odometry()
        odom.header.stamp = stamp
        odom.header.frame_id = self._map_frame          # map (inertial world origin)
        odom.child_frame_id = self._child_frame          # ground_truth_base_link

        t = robot_tf.transform.translation
        r = robot_tf.transform.rotation

        odom.pose.pose.position = Point(x=t.x, y=t.y, z=t.z)
        odom.pose.pose.orientation = Quaternion(x=r.x, y=r.y, z=r.z, w=r.w)

        # Covariance is zero for perfect ground truth
        odom.pose.covariance = [0.0] * 36

        # Twist is left as zero (velocity ground truth can be added later)
        odom.twist.covariance = [0.0] * 36

        self._odom_pub.publish(odom)

        # ── Publish PoseStamped ─────────────────────────────────────────────
        ps = PoseStamped()
        ps.header.stamp = stamp
        ps.header.frame_id = self._map_frame
        ps.pose.position = Point(x=t.x, y=t.y, z=t.z)
        ps.pose.orientation = Quaternion(x=r.x, y=r.y, z=r.z, w=r.w)

        self._pose_pub.publish(ps)

        self._last_stamp = stamp

    def _watchdog_cb(self) -> None:
        """Periodically log status so the user knows the node is alive."""
        if self._msg_count == 0:
            self.get_logger().warn(
                "No ground truth poses received yet.  "
                "Check that Gazebo is running and the bridge is active.\n"
                "  Expected bridge topic : /gz/dynamic_pose\n"
                f"  Expected child_frame  : {self._robot_name}"
            )
        else:
            self.get_logger().info(
                f"[ground_truth] Publishing OK — {self._msg_count} messages received."
            )
            self._msg_count = 0  # reset counter for next window


def main(args=None) -> None:
    rclpy.init(args=args)
    node = GroundTruthPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()
